update_test_checks: keep meta variables stable by default

nhaehnle opened the pull request:
There is a small sequence of commits here, the first one simply pre-commits a test. But really, you don't lose much by looking at the overall diff.

Prior to this change, running UTC on larger tests, especially tests with unnamed IR values, often resulted in a spuriously large diff because e.g. TMPnn variables in the CHECK lines were renumbered. This change attempts to reduce the diff by keeping those variable names the same.

There are cases in which this "drift" of variable names can end up being more confusing. The old behavior can be re-enabled with the `--reset-variable-names` command line argument.

The improvement may not be immediately apparent in the diff of this change. The point is that the diff of stable_ir_values.ll against stable_ir_values.ll.expected after this change is smaller.

Ideally, we'd also keep meta variables for "global" objects stable, e.g. for attributes (#nn) and metadata (!nn). However, that would require a much more substantial refactoring of how we generate check lines, so I left it for future work.

nhaehnle committed:
update_test_checks: precommit a test case

The test case demonstrates how meta variables are needlessly renamed,
making diffs harder to read.

nhaehnle committed:
update_test_checks: keep meta variables stable by default

Prior to this change, running UTC on larger tests, especially tests
with unnamed IR values, often resulted in a spuriously large diff
because e.g. TMPnn variables in the CHECK lines were renumbered. This
change attempts to reduce the diff by keeping those variable names the
same.

There are cases in which this "drift" of variable names can end up being
more confusing. The old behavior can be re-enabled with the
--reset-variable-names command line argument.

The improvement may not be immediately apparent in the diff of this change.
The point is that the diff of stable_ir_values.ll against
stable_ir_values.ll.expected after this change is smaller.

Ideally, we'd also keep meta variables for "global" objects stable, e.g.
for attributes (#nn) and metadata (!nn). However, that would require a
much more substantial refactoring of how we generate check lines, so I
left it for future work.

nhaehnle committed:
update_test_checks: simplify is_local_def_ir_value

The match argument is unused.

nhaehnle committed:
update_test_checks: simplify get_ir_regex

The match argument isn't used.

fhahn commented:
Not looked at the details yet, but great to see this!

jdoerfert reviewed (commented):
I haven't reviewed it (yet) but this sounds pretty cool.

jasilvanus commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1176,20 +1215,236 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+                for value in line.values:
+                    committed_names.add(new_mapping[value.var])
+
+        # Search for lines that are unique on both sides if we only consider
+        # variable names that have been committed.
```
For my understanding: This is an artificial restriction to make the algorithm simpler, correct? 

We rely on unique lines (ignoring non-committed variable names, so initially all variable names) dividing the input into hopefully small enough chunks that again hopefully contain unique pairs?

I imagine in corner cases (e.g. similar consecutive lines differing only in variable names) this could lead to us not re-using names although we could.

nhaehnle replied:
The point of this is to bound the runtime complexity: it keeps the number of edges in the matching graph linear, and so prevents the overall running time from becoming quadratic in the worst case.

This is apparently a fairly common trick in diff implementations.

It's not a significant simplification of the algorithm. The only gotcha is that if we allow non-unique matches, we need to be careful about how we sort the matching edges: they have to matched in *increasing* old line number first, and then in *decreasing* new line number. This isn't hard to see with the line sweeping intuition and the standard trick of applying an epsilon perturbation.

jasilvanus commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1176,20 +1215,236 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+        self.prefix = prefix
+
+
+class CheckLineInfo:
```
Maybe add an explanation how we represent lines and variables: `line` is the line with all variables replaced by `VARIABLE_TAG`, and variables are stored separately. This allows to match lines while ignoring some or all of variable names.

jasilvanus reviewed (commented):
Thanks, this is very useful. I often resort to custom diff tools trying to filter out test churn caused by renamed variables.

I think we should add more comments explaining how the algorithm works and why we are doing what we are doing (see my inline comments in that regard).

Could you please add a more complicated test that tests the recursion and crossing-free matching part? 
I suspect the existing test case is trivially resolved in the top level without any of this.

How strongly do you feel about enabling this by default (now)? How should this feature (and resetting of names) be integrated into development work flows?

I'm a bit worried about defaulting to behavior where checked-in test lines depends on history, but on the other hand not making this the default will mean it's often not used where applicable (or reviewers have to request it), and I'd expect most UTC uses to prefer names to be perserved, except for dedicated test normalization changes.

Independently, we may want to test this a bit more before enabling by default, but that is only a temporary consideration.

nikic commented:
If I may hijack this for an unrelated question, do you plan to submit your patch for block labels again? (At least I think that was yours -- I can't find it anymore.)

nhaehnle commented:
Thanks for the reminder. I imagine I'll still be digging myself out of other backlog items for a while, though...

nhaehnle commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1176,20 +1214,236 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+
+        backlinks = []
+        table = []
+        for _, new_idx in unique_matches:
+            ti = bisect.bisect_left(table, new_idx, key=lambda entry: entry[0])
+            if ti < len(table):
+                table[ti] = (new_idx, len(backlinks))
+            else:
+                table.append((new_idx, len(backlinks)))
+            if ti > 0:
+                backlinks.append(table[ti - 1][1])
+            else:
+                backlinks.append(None)
```
Yeah, you're missing something because this isn't *quite* a dynamic programming algorithm. It started out as one, but it's really more of a line-sweeping algorithm. I'm going to add a ~40 line comment explaining this.

nhaehnle commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1176,20 +1214,236 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+                    break
+
+                local_commits[new_value.var] = orig_value.var
+            else:
```
It's the one structured control flow construct that's still missing from most languages :)

nhaehnle commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1176,20 +1215,236 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+        for i, line in enumerate(orig_line_infos):
+            key = (
+                line.line.strip(),
+                tuple(
+                    value.var for value in line.values if value.var in committed_names
+                ),
```
You may be right. I will investigate.

nhaehnle commented:
I'm getting back to this now. I've rebased and addressed many of the review comments. I plan to test this a bit more now and perhaps simplify some of the logic around recursion unless I can find a compelling test case that benefits.

nhaehnle committed:
squash! update_test_checks: precommit a test case

nhaehnle committed:
squash! update_test_checks: keep meta variables stable by default

nhaehnle commented:
This PR is now ready for another round of reviews.

I've done a bunch more testing with real-world examples and ended up changing the high-level algorithmic approach somewhat to get better results. As a bonus, the core diffing algorithm is now in a separate function.

nhaehnle commented:
As for the question about whether to default this: I do think it should be the default, certainly in the long run. It could benefit from a bit more testing, but on the other hand the `--reset-variable-names` command-line option is a good way to bypass most of the new logic and can serve as a temporary workaround if a problem is indeed found.

jasilvanus reviewed (commented):
Thanks, factoring out the matching algorithm into a separate functions is a good change indeed. Also the detailed comments are very helpful.

Having the two similar recursion schemes independent of each other is a bit awkward though.

Is the inner recursion scheme even necessary if the outer one will recurse into intervals between matching edges anyways? It wouldn't be strictly the same as now, because name mapping happens in between, restricting some matching edges, but I think it would also work. Maybe even better, because we'd avoid matches that cannot be used due to name conflicts.

nhaehnle committed:
squash! update_test_checks: keep meta variables stable by default

- Remove inner recursion
- Add some more comments
- Minor fixups

nhaehnle commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1187,20 +1233,317 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+
+        multiple_patterns.sort(key=lambda pattern: len(pattern[0]) * len(pattern[1]))
+
+        for pattern in multiple_patterns:
+            if len(candidates) + len(pattern[0]) * len(pattern[1]) > 2 * (len(lhs) + len(rhs)):
```
Yes, I considered such options, but I also wanted to be able to write a reasonable test case that still hits the case where we drop candidates due to this check. If you feel really strongly about it (and can ideally support it with a concrete test case) I can look at it, but I'd prefer to just land something simpler for the first version.

jasilvanus replied:
I see, testability is an argument against just killing small cases with an additive constant. We can revisit later if feel the need to improve on real-world examples.

nhaehnle commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1187,20 +1233,317 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+        for _, rhs_idx in candidates:
+            candidate_idx = len(backlinks)
+            ti = bisect.bisect_left(table, rhs_idx, key=lambda entry: entry[0])
+            if ti < len(table):
+                table[ti] = (rhs_idx, candidate_idx)
+            else:
+                table.append((rhs_idx, candidate_idx))
```
There is actually a subtlety here that I'm recording in a comment, and I ended up not quite restating what you wrote here. I hope that's okay :)

jasilvanus replied:
Thanks, that's fine.

nhaehnle commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -1187,20 +1233,317 @@ def may_clash_with_default_check_prefix_name(check_prefix, var):
+        #    links back to the previous point on a best (longest) chain ending
+        #    at that point
+        #
+        # The core loop of the algorithm sweeps the line and updates the table
+        # and backlink structure for every point that we cross during the sweep.
+        # Therefore, the algorithm is trivially O(M log M) in the number of
+        # points. Since we only consider lines that are unique, it is log-linear
+        # in the problem size.
```
I dropped the inner recursion.

nhaehnle commented:
Thanks for taking another look.

Removing the inner recursion changes the outcome of one of the test cases, but not in a way that seems relevant to me in terms of quality. So I'm going to remove it to keep things a bit simpler.

And yes, with the outer recursion we don't have a near-linear time guarantee anymore. It's probably okay in this case?

jasilvanus reviewed (approved):
Thanks, this now looks good to me.

nhaehnle committed:
squash! update_test_checks: keep meta variables stable by default

Attempt to fix the Windows builder -- my guess is that a different
Python version doesn't accept this construct (though I don't understand
why it only fails Clang's update_cc_test_checks).

nhaehnle committed:
squash! update_test_checks: keep meta variables stable by default

nhaehnle commented:
Merged via CLI

Flakebi commented:
Commit fb02f9a (and parents) for future reference
